// prefix-parser/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatDiscordHelp = exports.wrap = exports.loremIpsum = exports.isNum = exports.isPopulated = exports.isVoid = exports.intArg = exports.assertGreaterThan = exports.assertPositive = exports.assert = void 0;
/** Throws an error is the given condition is not true. Used for validating developer inputs. */
function assert(check, message) {
    if (!check)
        throw Error(`Assert error: ${message}`);
}
exports.assert = assert;
/** Throws an error if the given argument value is not greater than or equal to 0. */
function assertPositive(argValue) {
    assert(argValue >= 0, `Argument with value '${argValue}' must be positive`);
}
exports.assertPositive = assertPositive;
/** Throws an error if the given argument value is not greater than or equal to the bound value. */
function assertGreaterThan(argValue, boundValue) {
    assert(argValue > boundValue, `Argument with value '${argValue}' must be greater than '${boundValue}'`);
}
exports.assertGreaterThan = assertGreaterThan;
/** Converts a numeric argument value to be a positive integer that (including 0). */
function intArg(num) {
    return Math.abs(Math.round(num));
}
exports.intArg = intArg;
function isVoid(val) {
    return (val === null || val === undefined || Number.isNaN(val));
}
exports.isVoid = isVoid;
function isPopulated(val) {
    return !isVoid(val);
}
exports.isPopulated = isPopulated;
function isNum(val) {
    return typeof val === 'number' && isPopulated(val);
}
exports.isNum = isNum;
function loremIpsum(wordCount) {
    const words = 'lorem ipsum dolor sit amet'.split(' ');
    return [...new Array(wordCount)].map((_, i) => words[i % words.length]).join(' ');
}
exports.loremIpsum = loremIpsum;
function wrap(message, char) {
    return char + message + char;
}
exports.wrap = wrap;
function formatDiscordHelp(title, usage, example) {
    return [
        title,
        '',
        '***Help***',
        usage,
        '',
        '***Example***',
        example,
    ].join('\n');
}
exports.formatDiscordHelp = formatDiscordHelp;
